UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API.UnitTests/Controllers/SubscriptionControllerTests.cs

```diff
@@ -15,40 +21,86 @@ public class SubscriptionControllerTests
+        private SubscriptionRequest _fakeSubscriptionRequest;
 
         [SetUp]
         public void Setup()
         {
+            _fakeD365PayloadDetails = GetD365Payload();
+            _fakeSubscriptionRequest = GetSubscriptionRequest();
+
             _fakeHttpContextAccessor = A.Fake<IHttpContextAccessor>();
             _fakeLogger = A.Fake<ILogger<SubscriptionController>>();
-            A.CallTo(() => _fakeHttpContextAccessor.HttpContext).Returns(new DefaultHttpContext());
+            _fakeSubscriptionService = A.Fake<ISubscriptionService>();
+
+            _controller = new SubscriptionController(_fakeHttpContextAccessor, _fakeLogger, _fakeSubscriptionService);
+        }
+
+        [Test] 
+        public async Task WhenPostInvalidNullPayload_ThenRecieveBadRequest()
+        {
+            D365Payload d365Payload = null;
+
+            var result = (BadRequestObjectResult)await _controller.Post(d365Payload);
+            var errors = (ErrorDescription)result.Value;
+
+            Assert.AreEqual(400, result.StatusCode);
+            Assert.AreEqual("Either body is null or malformed.", errors.Errors.Single().Description);
+        }
+
+        [Test] 
+        public async Task WhenPostInvalidNullInputParameters_ThenRecieveBadRequest()
+        {
+            var validationMessage = new ValidationFailure("InputParameters", "D365Payload InputParameters cannot be blank or null.")
+            {
+                ErrorCode = HttpStatusCode.BadRequest.ToString()
+            };
+
+            A.CallTo(() => _fakeSubscriptionService.ValidateD365PayloadRequest(A<D365Payload>.Ignored)).Returns(new ValidationResult(new List<ValidationFailure> { validationMessage }));
+
+            var result = (BadRequestObjectResult)await _controller.Post(_fakeD365PayloadDetails);
+            var errors = (ErrorDescription)result.Value;
 
-            _controller = new SubscriptionController(_fakeHttpContextAccessor, _fakeLogger);
+            Assert.AreEqual(400, result.StatusCode);
+            Assert.AreEqual("D365Payload InputParameters cannot be blank or null.", errors.Errors.Single().Description);
         }
 
         [Test]
         public async Task WhenPostValidPayload_ThenRecieveSuccessfulResponse()
         {
-            var result = (StatusCodeResult)await _controller.Post(GetD365Payload());
+            A.CallTo(() => _fakeSubscriptionService.ValidateD365PayloadRequest(A<D365Payload>.Ignored)).Returns(new ValidationResult(new List<ValidationFailure>()));
+
+            A.CallTo(() => _fakeSubscriptionService.ConvertToSubscriptionRequestModel(A<D365Payload>.Ignored)).Returns(_fakeSubscriptionRequest);
+            
+            var result = (StatusCodeResult)await _controller.Post(_fakeD365PayloadDetails);
+
             Assert.AreEqual(StatusCodes.Status202Accepted, result.StatusCode);
         }
 
-        private static D365Payload GetD365Payload()
+        private D365Payload GetD365Payload()
+        {
```

Comment:
Please keep it static - private static D365Payload GetD365Payload()

Reply:
Done required modification.